# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2004/06/15 12:06:00-05:00 akpm@osdl.org 
#   [PATCH] drivers/scsi/megaraid.c: user/kernel pointer bugs
#   
#   From: "Robert T. Johnson" <rtjohnso@eecs.berkeley.edu>
#   
#   Since arg is a user pointer, so are uioc_mimd and uiocp, and hence umc is a
#   user pointer.  Thus reading umc->xferaddr requires dereferencing a user
#   pointer, which isn't safe.
#   
#   Signed-off-by: Andrew Morton <akpm@osdl.org>
#   Signed-off-by: James Bottomley <James.Bottomley@SteelEye.com>
# 
# drivers/scsi/megaraid.c
#   2004/06/10 08:19:19-05:00 akpm@osdl.org +4 -2
#   drivers/scsi/megaraid.c: user/kernel pointer bugs
# 
diff -Nru a/drivers/scsi/megaraid.c b/drivers/scsi/megaraid.c
--- a/drivers/scsi/megaraid-old.c	2004-06-22 12:21:30 -07:00
+++ b/drivers/scsi/megaraid-old.c	2004-06-22 12:21:30 -07:00
@@ -3815,7 +3815,8 @@
 
 			umc = MBOX_P(uiocp);
 
-			upthru = (mega_passthru *)umc->xferaddr;
+			if (get_user(upthru, (mega_passthru **)&umc->xferaddr))
+				return (-EFAULT);
 
 			if( put_user(mc->status, (u8 *)&upthru->scsistatus) )
 				return (-EFAULT);
@@ -3831,7 +3832,8 @@
 
 			umc = (megacmd_t *)uioc_mimd->mbox;
 
-			upthru = (mega_passthru *)umc->xferaddr;
+			if (get_user(upthru, (mega_passthru **)&umc->xferaddr))
+				return (-EFAULT);
 
 			if( put_user(mc->status, (u8 *)&upthru->scsistatus) )
 				return (-EFAULT);
